pkg/pool-utils/test/BasePoolController.test.ts

```diff
@@ -0,0 +1,236 @@
+let other: SignerWithAddress;
+let governance: SignerWithAddress;
+let pool: WeightedPool;
+let allTokens: TokenList;
+let vault: Vault;
+let poolController: Contract;
+
+before('setup signers', async () => {
+  [admin, other, governance] = await ethers.getSigners();
+});
+
+sharedBeforeEach('deploy Vault and tokens', async () => {
+  vault = await Vault.create({
+    admin,
+    pauseWindowDuration: PAUSE_WINDOW_DURATION,
+    bufferPeriodDuration: BUFFER_PERIOD_DURATION,
+  });
+
+  allTokens = await TokenList.create(['MKR', 'DAI', 'SNX', 'BAT'], { sorted: true });
+  await allTokens.mint({ to: admin, amount: fp(100) });
+});
+
+describe('BasePoolController', function () {
+  const NEW_SWAP_FEE = fp(0.05);
+  const NEXT_SWAP_FEE = fp(0.005);
+
+  sharedBeforeEach('deploy controller and pool', async () => {
+    poolController = await deploy('BasePoolController');
+
+    const params = {
+      vault,
+      tokens: allTokens,
+      weights: WEIGHTS,
+      owner: poolController.address,
+      swapFeePercentage: POOL_SWAP_FEE_PERCENTAGE,
+      poolType: WeightedPoolType.WEIGHTED_POOL,
+      swapEnabledOnStart: false,
+    };
+    pool = await WeightedPool.create(params);
+  });
+
+  context('pool controller not initialized', () => {
+    it('creates the pool', async () => {
+      expect(await pool.getOwner()).to.equal(poolController.address);
+    });
+
+    it('sets up the pool controller', async () => {
+      expect(await poolController.owner()).to.equal(admin.address);
+    });
+
+    it('cannot call functions before initialization', async () => {
+      await expect(poolController.connect(admin).setSwapFeePercentage(NEW_SWAP_FEE)).to.be.revertedWith(
+        'UNINITIALIZED'
+      );
+    });
+
+    it('has no initial metadata', async () => {
+      expect(await poolController.getMetadata()).to.equal('0x');
+    });
+
+    it('owner can set metadata', async () => {
+      const tx = await poolController.connect(admin).updateMetadata(METADATA);
+      expectEvent.inReceipt(await tx.wait(), 'MetadataUpdated', {
+        metadata: METADATA,
+      });
+
+      expect(await poolController.getMetadata()).to.equal(METADATA);
+    });
+
+    it('non-owner cannot set metadata', async () => {
+      await expect(poolController.connect(other).updateMetadata(METADATA)).to.be.revertedWith('CALLER_IS_NOT_OWNER');
+    });
+  });
+
+  context('pool controller is initialized', () => {
+    sharedBeforeEach('initialize pool controller', async () => {
+      await poolController.bindPool(pool.address);
+    });
+
+    it('owner can set the swap fee', async () => {
+      await poolController.connect(admin).setSwapFeePercentage(NEW_SWAP_FEE);
+
+      expect(await pool.getSwapFeePercentage()).to.equal(NEW_SWAP_FEE);
+    });
+
+    it('enforces owner permissions', async () => {
+      await expect(poolController.connect(other).setSwapFeePercentage(NEW_SWAP_FEE)).to.be.revertedWith(
+        'CALLER_IS_NOT_OWNER'
+      );
+    });
```

Comment:
These tests could use with a bit more structuring. 'enforces owner permission' is only checking that  `setSwapFeePercentage` is `onlyOwner`, but doesn't check anything about the other functions. And since everything is in the same scope, it is hard to review how thorough our tests are.

We could have something like this:

```
describe set swap fee
  it lets the owner set the swap fee
  it reverts if a non-owner attempts to set the swap fee
describe set asset manager thing
 (repeat the above)
```

This could even be abstracted away into a function (`itEnsuresTheSenderIsTheOwner`) that we call in each test.